systemd-hosts.d.service

```diff
@@ -3,13 +3,66 @@ Description=Keep your hosts records in logically separated files
 After=local-fs.target
 
 [Service]
+# Security
+ProtectSystem=yes
+ProtectHome=yes
+PrivateDevices=yes
+ProtectKernelTunables=yes
+ProtectKernelModules=yes
+ProtectControlGroups=yes
+ProtectKernelLogs=yes
+SystemCallFilter=@system-service
+SystemCallErrorNumber=EPERM
+NoNewPrivileges=yes
+PrivateTmp=yes
+LockPersonality=yes
+MemoryDenyWriteExecute=yes
+ProtectClock=yes
+RestrictNamespaces=yes
+RestrictRealtime=yes
+RestrictSUIDSGID=yes
+RuntimeDirectory=systemd/hosts.d
+RuntimeDirectoryPreserve=yes
+
+# Service
 Type=oneshot
+
+# Environment
 Environment="HOSTS_FILE=/etc/hosts"
 Environment="HOSTS_DIRECTORY=/etc/hosts.d"
-ExecStart=/bin/sh -c 'echo -e "## Generated by systemd-hosts.d ##\\n" > $HOSTS_FILE'
-ExecStart=/bin/sh -c 'echo -e "## All changes in this file will be ignored ##\\n" >> $HOSTS_FILE'
-ExecStart=/bin/sh -c 'cat $HOSTS_DIRECTORY/*.conf | grep \'^[[:blank:]]*[^[:blank:]#;]\' >> $HOSTS_FILE'
-ExecStart=/bin/chmod ugo=r $HOSTS_FILE
+# Loosely match IPv4 or IPv6 followed by hostname and hostalias(s)
+Environment="HOSTS_REGEX_BASE=^\\([0-9]\\{1,3\\}\\.\\|[0-9a-fA-F]\\{1,4\\}:\\|::\\).\\+\\?\\(\\s\\+[a-zA-Z0-9.-]\\{1,\\}\\)\\{1,\\}$"
+Environment="HOSTS_REGEX_COMMENT=^#.*$"
+# Override with HOSTS_ENABLE_COMMENT=yes to include comments from the source conf files
+Environment="HOSTS_ENABLE_COMMENT=no"
+# Override with HOSTS_ENABLE_BACKUP=no to disable hosts file backup creation
+Environment="HOSTS_ENABLE_BACKUP=yes"
+
+# Condition - validate that matching regex is found in the source before overwriting $HOSTS_FILE
+ExecCondition=/bin/sh -c 'grep "$HOSTS_REGEX_BASE" "$HOSTS_DIRECTORY"/*.conf >/dev/null'
```

Comment:
Good observation. We indeed want to have a fail safe in the case the package was improperly installed. I think the condition could be simplified to pass if there is just any matching file in the hosts directory.

Reply:
I'd experienced the exact scenario on one host although it was "properly" installed. My existing `/etc/hosts.d/hosts.conf` (generated by copying `/etc/hosts`) was effectively empty since it had only comments. This causes the service to fail entirely which is undesirable for a configuration that's valid with `systemd-resolved` and other local resolver.